torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1733,6 +1736,82 @@ struct to_ir {
     emitLoopCommon(stmt.range(), emit_body, nullptr, {}, cond);
   }
 
+  void emitWith(const With& stmt) {
+    auto targets = stmt.targets();
+    // Keep a stack of entered objects so they can be exited
+    // in the right order.
+    std::stack<Value*> entered;
+    Block* current_block = nullptr;
+
+    for (const auto& target : targets) {
+      if (target.kind() == TK_WITH_ITEM_AS) {
+        // This item is of the form "e as i".
+        auto withItem = WithItemAs(target);
+        Expr e = withItem.target();
+        Var i = withItem.var();
+        auto v = emitSugaredExpr(e, 1);
+        auto* n = graph->create(prim::Enter);
+        graph->insertNode(n);
+        current_block = n->owningBlock();
+        // Treat each with item as an assignment i = e.__enter__().
+        auto* rhs = v->asValue(e.range(), method);
+        n->addInput(rhs);
+        entered.push(rhs);
+        auto rhsClass = rhs->type()->expect<ClassType>();
+
+        if (!rhsClass) {
+          throw ErrorReport(e.range())
+              << "With item expression does not return a class type";
+        }
+
+        auto* enterMethod = rhsClass->getMethod("__enter__");
+
+        if (!enterMethod) {
+          throw ErrorReport(e.range())
+              << "Object returned by with item expression does not define an __enter__ method";
+        }
+
+        // Set the output of the enter node to be the return type of __enter__.
+        n->output(0)->setType(rhsClass->getMethod("__enter__")
```

Comment:
I think we should be doing some checking of the `__enter__` and `__exit__` schema to check that they are of an expected form.

Reply:
Here are the schemas from one of my tests:
```
__enter__(__torch__.jit.test_with.Context self) -> (int)
__exit__(__torch__.jit.test_with.Context self, Tensor type, Tensor value, Tensor tb) -> (None)
```
For both methods, the first arg must be `self` because these are both class methods, and they are fetched from the `ClassType` object corresponding to the class they belong to. So checking that the first arg is `self` seems pointless.

In the case of `__enter__`, the return type is `int` because that's how I wrote the example. That' s not always the case. 

In the case of `__exit__`, we don't actually use the last three args, but they have to be present. So I think it would be good to check that `__exit__` takes 4 args.